### ISSUE TITLE
Adding flow control to Serenity-js specs

### ISSUE THREAD

**Issue by nbarrett** [jan-molak/serenity-js#155]
As an keen user of the [java version of serenity](https://github.com/serenity-bdd/serenity-core), I frequently use the [ConditionalPerformable](https://github.com/serenity-bdd/serenity-core/blob/master/serenity-screenplay/src/main/java/net/serenitybdd/screenplay/conditions/ConditionalPerformable.java) in order to add if/then/else flow control to the execution. For instance I'd like to do something like this:

```
      actor.attemptsTo(
           Check.whether(ContactsQuestions.errorShown())
              .andIfSo(
                    Click.on(ContactsTargets.addContact),
                    Enter.theValue("nick").into(ContactsTargets.firstName),
                    Enter.theValue("barrett").into(ContactsTargets.lastName),
                ).otherwise(                    
                    Click.on(ContactsTargets.resetForm)));
```
It looks like the concept of conditional flow doesn't exist yet in serenity-js so is there a way to implement this myself using a hand-written `Interaction`?

**Comment by jan-molak** [comment]
Yeah, I don't see why not, it would be something along those lines:

```typescript
const Check = {
  whether: (question: Question<PromiseLike<boolean>>) => ({
    andIfSo: (activitiesWhenTrue: Activity[]) => ({
      otherwise: (activitiesWhenFalse: Activity[]) => Interaction.where(
        `#actor makes a decision based on ${question}`,
        actor => question.answeredBy(actor).
                          then(result => result 
                                         ? actor.attemptsTo(...activitiesWhenTrue)
                                         : actor.attemptsTo(...activitiesWhenFalse)
                          ),                     
    }),
  }),
};
```

Does this help?

**Comment by nbarrett** [comment]
Hi @jan-molak - yes that's looking great thanks! Only thing is that my IntelliJ intellisense is only offering .abilityTo, .toSee and .whoCan on the actor at the bottom there...

![image](https://user-images.githubusercontent.com/626941/35553594-76cd87f0-0590-11e8-9013-b6c1190a5b91.png)

**Comment by jan-molak** [comment]
Ah, drats :) Yes, that's because `Interaction` assumes a more narrows interface (`UsesAbilities`) than what you'd need for that (`PerformsTasks`).

In this case, you might need to go with something akin to ConditionalPerformable.

**Comment by nbarrett** [comment]
Ah, well in that case I'm running out of ideas! I wouldn't have much of a clue how to do that, hence raising this ticket 😊. Any pointers on approach much appreciated please!

**Comment by nbarrett** [comment]
Okay, I braved it and had a go at fleshing something out and came up with this (which all seems to compile 😼 :

```
import { Actor, AnswersQuestions, PerformsTasks, UsesAbilities } from './actor';
import { Question } from './question';
import { Activity, Interaction } from './activities';

export class SilentPerformable implements Interaction {
    performAs(actor: PerformsTasks | UsesAbilities | AnswersQuestions): PromiseLike<void> {
        return undefined;
    }
}

export abstract class ConditionalPerformable implements Activity {

    protected outcomeToPerform: { [key: string]: Activity[]; } = {};

    constructor() {
        this.outcomeToPerform['true'] = [new SilentPerformable()];
        this.outcomeToPerform['false'] = [new SilentPerformable()];
    }

    andIfSo = (...activities: Activity[]) => this.outcomeToPerform['true'] = activities;
    otherwise = (...activities: Activity[]) => this.outcomeToPerform['false'] = activities;

    abstract evaluatedConditionFor(actor: Actor): boolean;

    performAs(actor: PerformsTasks): PromiseLike<void> {
        return actor.attemptsTo(...this.outcomeToPerform[this.evaluatedConditionFor(actor).toString()]);
    }
}

class ConditionalPerformableOnBoolean extends ConditionalPerformable {

    constructor(private condition: boolean) {
        super();
    }

    evaluatedConditionFor(actor: Actor) {
        return this.condition;
    }
}

class ConditionalPerformableOnQuestion extends ConditionalPerformable {

    constructor(private condition: Question<boolean>) {
        super();
    }

    evaluatedConditionFor(actor: Actor) {
        return this.condition.answeredBy(actor);
    }
}
```
Is this the direction I should be going in? I'll have a go and doing some specs so that I can get my code through the molak-gate and we'll take it from here?

**Comment by nbarrett** [comment]
Hey @jan-molak - see above PR #159!

**Comment by jan-molak** [comment]
Fixed in [3883ece9](https://github.com/jan-molak/serenity-js/commit/3883ece9819f6cc419edfc4a8e6b64fec0d78f2a)
